### PR TITLE
BZNG canonical set representation — PROD-1296

### PR DESCRIPTION
**Pull request by nwt**
Define and enforce a normal form for BZNG sets.  This makes it possible
to compare sets for equality by comparing their BZNG byte
sequences.  It also fixes incorrect zng.Value.ContainerLength results
caused by duplicate elements.

Tests are coming.

### CODE REVIEW

File: zcode/builder.go [hunk 1]
```diff
@@ -31,11 +33,15 @@ func (b *Builder) BeginContainer() {
 }
 
 // EndContainer closes the most recently opened container.  It panics if the
-// receiver has no open container.
-func (b *Builder) EndContainer() {
+// receiver has no open container.  isSet must be true if and only if the
+// container is a set.
+func (b *Builder) EndContainer(isSet bool) {
```
Comment (nwt) [inline zcode/builder.go]: I think adding this boolean makes for a worse interface, but I'm not sure what else to do. I'd have preferred `EndContainer(typ *zng.Type)`, but that causes a circular import dependency.

Comment (henridf) [reply zcode/builder.go]: This is just cosmetic, but how about renaming `isSet` to `normalize` ?  It seems like the builder API is lower-level and doesn't know about zng types.

Along those lines, and related to your comment about the interface, another option would be to leave the `BeginContainer()` and `EndContainer()` unchanged, and having a separate `NormalizeContainer()` call that a client can call any time while in a container (with the common and expected case being to call it just before calling `EndContainer()`. Just an idea, entirely your call.

**Comment by nwt** [comment]
Tests have arrived.

**Comment by henridf** [comment]
> We've done such a good job with the type-independent zcode.Bytes format and now you are passing in type info (isSet bool). It seems the spec should require normal format of the set value and the only place where you need to fix it up would be the zeek reader.

@mccanne I had the same thought about zng types into zcode.Bytes (comment above), but under what encoding would you do the normalization if it moves out? a normalization in zng wouldn't be the same as in bzng...

File: zcode/builder.go (hunk 1 above)
Comment (alfred-landrum) [reply zcode/builder.go]: How about this option:
- keep EndContainer() as is before the PR.
- add an additional call, `EndContainerTransform(func transform(zcode.Bytes) zcode.Bytes))`, that does the same as EndContainer, but calls the passed in `transform` with the container bytes, and replaces them with the return result?

**Comment by nwt** [comment]
> We've done such a good job with the type-independent zcode.Bytes format and now you are passing in type info (isSet bool). It seems the spec should require normal format of the set value and the only place where you need to fix it up would be the zeek reader.

I wound up here because I thought we should normalize sets in all non-bzng readers. I still do because set normalization seems like a troublesome requirement to impose on non-bzng writers, but that's a position I'd happy to be talked out of.

We can certainly avoid doing set normalization in zcode, though. Let me give that a shot.

**Comment by nwt** [comment]
> I had the same thought about adding zng types into the zcode API (comment above), but under what encoding would you do the normalization if it moves out? a normalization in zng wouldn't be the same as in bzng...

One approach would be to accumulate set values in a []zcode.Bytes rather than in the zcode.Builder, then normalize the slice, and finally append the values. It's probably less efficient, but probably not significantly.

**Comment by nwt** [comment]
> One approach would be to accumulate set values in a []zcode.Bytes rather than in the zcode.Builder, then normalize the slice, and finally append the values. It's probably less efficient, but probably not significantly.

Doing this for all non-bzng readers is turning into big diff, so I'm pressing pause on that.

@henridf, @mccanne: Want to weight in on whether I try @alfred-landrum's [EndContainerTransform approach](https://github.com/brimsec/zq/pull/282#discussion_r377157532) or, alternatively, just drop set normalization for the non-zeek readers?

**Comment by henridf** [comment]
@nwt I'd lean towards your proposed approach of accumulating set values in a []zcode.Bytes, and have normalization be done by a new `zng.NewSet()`, following our existing model for primitive types. And since that sounds like too big of a change for now, I'd start off with doing this for zeek only, and come back to the other readers later.

**Comment by henridf** [comment]
Separately from what is implemented here, I'd propose that the spec requires normalization for sets in their bzng encoding only. A bzng-encoding of set with repeated or non-sorted elements is invalid. However, non-normalized sets are valid in all other formats. For example `[[1.0;1.000]]` is a redundant yet valid representation of a single element `set[double]` in zng.

This is in line with how non-bzing formats admit non-normalized representations of primitive values, such as admitting "1.000" as the encoding of the double "1.0".

File: zcode/builder.go (hunk 1 above)
Comment (nwt) [reply zcode/builder.go]: I went with a variation on @alfred-landrum's suggestion that I found more ergonomic:
* Keep `func (b *Builder) EndContainer()`
* Add `func (b *Builder) TransformContainer(transform func(Bytes) Bytes)`

**Comment by nwt** [comment]
I've updated this to avoid infecting zcode with type information and think it's ready for another look.

File: zcode/iter.go [hunk 2]
```diff
@@ -31,3 +31,19 @@ func (i *Iter) Next() (Bytes, bool, error) {
 	*i = (*i)[end:]
 	return Bytes(val), tagIsContainer(u64), nil
 }
+
+// NextTagAndBody returns the next value as a slice containing the value's
+// undecoded tag followed by its body along with a boolean this is true if the
```
[the remark this answers is not in the document]
Comment (nwt) [reply zcode/iter.go]: 👍

File: zng/set.go [hunk 3]
```diff
@@ -96,3 +98,31 @@ func (t *TypeSet) Marshal(zv zcode.Bytes) (interface{}, error) {
 	}
 	return vals, nil
 }
+
+// NormalizeSet interprets zv as set body and returns a set body that has been
```
[the remark this answers is not in the document]
Comment (nwt) [reply zng/set.go]: 👍

**Pull request merged**